packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -1725,6 +1726,33 @@ function validateAttributes(
     }
 }
 
+function validateSlotAttribute(
+    ctx: ParserCtx,
+    parsedAttr: ParsedAttribute,
+    parentNode: ParentNode,
+    element: BaseElement
+): void {
+    const slotAttr = parsedAttr.get('slot');
+
+    if (!slotAttr) {
+        return;
+    }
+
+    if (
+        ast.isElement(parentNode) ||
+        ast.isSlot(parentNode) ||
+        ast.isForEach(parentNode) ||
+        ast.isForOf(parentNode)
+    ) {
+        ctx.warnOnNode(ParserDiagnostics.IGNORED_SLOT_ATTRIBUTE_IN_CHILD, slotAttr, [
+            `<${element.name}>`,
+            ast.isElement(parentNode) || ast.isSlot(parentNode)
+                ? `<${parentNode.name}>`
+                : 'iteration block',
+        ]);
+    }
```

Comment:
Checking the parent node is a good start, but technically we'd need to walk up the tree to check if there are _any_ unexpected nodes between this node and its nearest ancestor Component (aka LWC Component) or ExternalComponent (i.e. `lwc:external`). E.g. this is invalid:

```html
<x-foo>
  <div> <!-- Invalid grandparent -->
    <template lwc:if={isTrue}> <!-- Valid parent -->
      <span slot=foo></span> <!-- Invalid! -->
    </template>
  </div>
</x-foo>
```

Reply:
I pushed an implementation and test for this.